File: problems/15.matrix/ts/bfs-traverse.ts
const directions = [
    [-1, 0],
    [0, +1],
    [+1, 0],
    [0, -1],
];

export function bfsTraverse<T>(matrix: T[][]) {
    const seen: boolean[][] = Array.from(
        { length: matrix.length },
        (_, index) => {
            return Array.from<boolean>({ length: matrix[index].length }).fill(
                false
            );
        }
    );
    let initialRow = Math.ceil((matrix.length - 1) / 2);
    let initialCol = Math.ceil((matrix[0].length - 1) / 2);
    let queue = [[initialRow, initialCol]];
    let result = [matrix[initialRow][initialCol]];
    seen[initialRow][initialCol] = true;

    while (queue.length > 0) {
        const [row, col] = queue.shift() as [number, number];

        for (let i = 0; i < directions.length; ++i) {
            let direction = directions[i];
            if (
                matrix[row + direction[0]] &&
                matrix[row + direction[0]][col + direction[1]] &&
                !seen[row + direction[0]][col + direction[1]]
            ) {
                const element = matrix[row + direction[0]][col + direction[1]];
                result.push(element);
                queue.push([row + direction[0], col + direction[1]]);
                seen[row + direction[0]][col + direction[1]] = true;
            }
        }
    }
    return result;
}

File: problems/15.matrix/golang/bfs_traverse.go
package main

var directions = [][]int{
	{-1, 0},
	{0, +1},
	{+1, 0},
	{0, -1},
}

func getInitialSeen(rowLength, colLength int) [][]bool {
	var seen [][]bool

	for i := 0; i < rowLength; i++ {
		for j := 0; j < colLength; j++ {
			seen[i][j] = false
		}
	}

	return seen
}

func BfsTraverseMatrix(matrix [][]int, initialRow, initialCol int, result []int) {
	//

	seen := getInitialSeen(len(matrix), len(matrix[0]))
	queue := [][]int{{initialRow, initialCol}}

	for len(queue) > 0 {
		coordinate := queue[0]
		queue = queue[1:]
		row := coordinate[0]
		col := coordinate[1]

		for i := 0; i < len(directions); i++ {
			direction := directions[i]
			tempRow := row + direction[0]
			tempCol := col + direction[1]

			////////////////////// we'
		}
	}

}
